[WPE] WPE Platform: Add WPESettings API

TingPing opened the pull request:
#### f076fdf8692a441dce2b82f193b8976d4436d424
<pre>
[WPE] WPE Platform: Add WPESettings API
<a href="https://bugs.webkit.org/show_bug.cgi?id=280618">https://bugs.webkit.org/show_bug.cgi?id=280618</a>

Reviewed by Carlos Garcia Campos.

This adds an API to store generic settings similar to GSettings.

Each setting has a known type and default value.

These settings can be saved and loaded from key files.

* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_settings):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/WPESettings.cpp: Added.
(SettingEntry::SettingEntry):
(SettingEntry::value const):
(wpe_settings_get_default):
(makeKeyPath):
(wpe_settings_register):
(wpe_settings_load_from_keyfile):
(wpe_settings_save_to_keyfile):
(wpe_settings_set_value):
(wpe_settings_get_value):
* Source/WebKit/WPEPlatform/wpe/WPESettings.h: Added.
* Tools/Scripts/run-wpe-tests:
(WPETestRunner):
(WPETestRunner.is_glib_test):
* Tools/TestWebKitAPI/Tests/WPEPlatform/TestSettings.cpp: Added.
(testSettingsRegistration):
(testSettingsSaveKeyFiles):
(testSettingsLoadKeyFiles):
(testSettingsChangedSignal):
(testSettingsHelperAPI):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/285702@main">https://commits.webkit.org/285702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d881aa2be6ab9f92e38f9c4944545d7d656731a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73404 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52833 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77658 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24642 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75519 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/61961 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/616 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57672 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16124 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76471 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63151 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38085 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20633 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22973 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66179 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79287 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/719 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/223 "Found 2 new test failures: http/tests/paymentrequest/ApplePayModifier-automaticReloadPaymentRequest.https.html http/tests/ssl/applepay/ApplePayPaymentDetailsModifier.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66083 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/861 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63163 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65363 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9204 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7384 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11333 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/683 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3420 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/713 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/697 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/715 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    if (!settings)
+        settings = static_cast<WPESettings*>(g_object_new(WPE_TYPE_SETTINGS, nullptr));
+    return settings;
+}
```
I think we want settings to be per display, not global. We might have a default method, but it would return the settings of the primary display.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    if (settingsObject->priv->settings.contains(key)) {
+        g_warn_if_reached();
+        return;
+    }
```
Since you added an error quark, maybe it would be better to make this function boolean and fill a GError in this case.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+ * to watch a specific key, e.g. `changed::/wpe-platform/fonts/font-hinting`.
+ */
+struct _WPESettingsPrivate {
+    WTF::HashMap<CString, SettingEntry> settings;
```
```suggestion
    HashMap<CString, SettingEntry> settings;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    settingsObject->priv->settings.add(key, SettingEntry {
+        nullptr,
+        g_variant_ref_sink(defaultValue),
+        GUniquePtr<GVariantType>(g_variant_type_copy(type))
```
I think we could add a constructor for the Entry like `Entry(const GVariantType* type, GRefPtr<GVariant>&& defaultValue)`

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    g_return_val_if_fail(keyFile, FALSE);
+
+    GUniquePtr<char*> groups(g_key_file_get_groups(keyFile, nullptr));
+    for (uint i = 0; groups.get()[i]; i++) {
```
```suggestion
    for (unsigned i = 0; groups.get()[i]; i++) {
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+
+        GUniquePtr<char*> keys(g_key_file_get_keys(keyFile, group, nullptr, nullptr));
+
+        for (uint k = 0; keys.get()[k]; k++) {
```
```suggestion
        for (unsigned k = 0; keys.get()[k]; k++) {
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+                continue;
+
+            auto path = makeKeyPath(group, key);
+            auto iter = settingsObject->priv->settings.find(path.data());
```
```suggestion
            auto iter = settingsObject->priv->settings.find(path);
```

TingPing replied:
Not sure why but it fails to find the `CString`...

carlosgcampos replied:
That's weird, because the key is a CString, passing its data is actually creating another CString for the same data to find it.

carlosgcampos replied:
Could you check the length of the CString compared to strlen(path.data())?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+
+            GRefPtr<GVariant> parsedValue = adoptGRef(g_variant_parse(iter->value.type.get(), value.get(), nullptr, nullptr, error));
+            if (!parsedValue)
+                return FALSE;
```
We should fill the error in this case.

TingPing replied:
`g_variant_parse()` sets `error` itself.

carlosgcampos replied:
Right, but since we have a specific error quark, I think we should use it for any error in this api. So, I would build a `WPE_SETTINGS_ERROR` with the info from the variant error in the message.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+
+    for (auto& kvPair : settingsObject->priv->settings) {
+        const auto& key = kvPair.key;
+        const auto& entry = kvPair.value;
```
```suggestion
    for (auto& [key, entry] : settingsObject->priv->settings) {
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    g_return_val_if_fail(WPE_IS_SETTINGS(settingsObject), FALSE);
+    g_return_val_if_fail(!error || !*error, FALSE);
+    g_return_val_if_fail(key && *key == '/', FALSE);
+    g_return_val_if_fail(value, FALSE);
```
Should we allow passing nullptr as value to reset to its default?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+ * @value: the value to set
+ * @error: return location for a #GError, or %NULL
+ *
+ * To set a value @key must have been registered and @value must be of the correct type.
```
We should document we are sinking the floating ref.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    if (iter == settingsObject->priv->settings.end()) {
+        g_warn_if_reached();
+        return nullptr;
+    }
```
This should fill a GError too, like the setter.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,342 @@
+    { \
+        GVariant* value = wpe_settings_get_value(settings, key); \
+        if (!value)\
+            return retval; \
```
This should also fill the GError

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.h:
```diff
@@ -0,0 +1,97 @@
+
+WPE_API WPESettings *wpe_settings_get_default();
+
+WPE_API void wpe_settings_register(WPESettings *settings, const char *key, const GVariantType *type, GVariant *default_value);
```
This should be formatted like all other public headers

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WPEPlatform/TestSettings.cpp:
```diff
@@ -0,0 +1,211 @@
+public:
+    MAKE_GLIB_TEST_FIXTURE(SettingsTest);
+
+};
```
You don't need this, you can use Test directly

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WPEPlatform/TestSettings.cpp:
```diff
@@ -0,0 +1,211 @@
+
+static void testSettingsRegistration(SettingsTest*, gconstpointer)
+{
+    GRefPtr<WPESettings> settings = adoptGRef(static_cast<WPESettings*>(g_object_new(WPE_TYPE_SETTINGS, nullptr)));
```
```suggestion
    GRefPtr<WPESettings> settings = adoptGRef(WPE_SETTINGS(g_object_new(WPE_TYPE_SETTINGS, nullptr)));
```

carlosgcampos reviewed (commented):
I like this a lot, and you are even adding API tests. The only thing missing is API for applications to override settings. Platforms can set their settings from system settings, for example, but the browser might want to set a different value. GTK uses a different `source` when the settings is changed internally from system settings and when set by apps using `g_object_set`. We need specific api for that. We could add wpe_settings_override_value(), or add a boolean parameter to current settings to tell whether to override the value. And document that platform implementationsd should never override. Maybe override is not the right word, though, but the idea is that once a value has been set by apps it shouldn't change when the corresponding system setting changes.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+    GRefPtr<GVariant> setValue;
+    GRefPtr<GVariant> defaultValue;
+    GUniquePtr<GVariantType> type;
+    bool setByApplication { false };
```
Move this block after the methods

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+
+    SettingEntry() = default;
+
+    explicit SettingEntry(GRefPtr<GVariant>&& defaultValue, GUniquePtr<GVariantType>&& type)
```
```suggestion
    SettingEntry(GRefPtr<GVariant>&& defaultValue, GUniquePtr<GVariantType>&& type)
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+ */
+WPESettings* wpe_settings_get_default()
+{
+    auto* display = wpe_display_get_default();
```
This would have the side effect of creating a display, I would either use the primary display or remove this get_default() helper.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+ * Registers a key to be used for settings. The key must begin with `/wpe-platform/` and
+ * not end with `/`. e.g. `/wpe-platform/fonts/font-hinting`.
+ *
+ * It is an error to call this function with a key that has already been registered.
```
We should also document that this is expected to be used only by platform implementations, applications are not expected to register settings.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+        return FALSE;
+    }
+
+    settingsObject->priv->settings.add(key, SettingEntry(g_variant_ref_sink(defaultValue), GUniquePtr<GVariantType>(g_variant_type_copy(type))));
```
```suggestion
    settingsObject->priv->settings.add(key, SettingEntry(defaultValue, GUniquePtr<GVariantType>(g_variant_type_copy(type))));
```

carlosgcampos replied:
GrefPtr already calls ref_sink on construction.

TingPing replied:
Doesn't appear to, tests crash on deref.

I think my usage is fine:

```cpp
    GRefPtr<GVariant> someValue = adoptGRef(g_variant_new_int32(42));
    wpe_settings_register(settings.get(), "/wpe-platform/a-new-key", G_VARIANT_TYPE_INT32, someValue.get(), nullptr);
```

carlosgcampos replied:
I don't know what the problem is, but that usage is not right for variants (or any other initially unowned type).

carlosgcampos replied:
The crash is fixed by not adopting floating references in test.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+ * @error: return location for a #GError, or %NULL
+ *
+ * If @override_platform is %TRUE, then the value will not be overwritten by the platform.
+ * This is intended for applications built on top of the platform.
```
I would include here that this should always be FALSE when called from platform implementations

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+ * @settings: a #WPESettings
+ * @key: the key to set
+ * @value: (transfer floating) (nullable): the value to set or %NULL
+ * @override_platform: whether to override the platform value
```
Well, then platform thing is the intention of this api, but this could be just override, and document that this overrides the existing value.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    if (iter->value.setByApplication && !overridePlatform)
+        return TRUE;
```
So, we just check if there's a value already set, and if override is FALSE we return.

TingPing replied:
Yes. If it was overridden before and this isn't an override then its a platform value to be ignored.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,374 @@
+
+#define g_variant_get_string(str) g_variant_get_string(str, nullptr)
+#define WPE_SETTINGS_HELPER_API(type, name, gvariant_type, retval) \
+    type wpe_settings_get_##name(WPESettings* settings, const char* key, GError** error) \
```
I've just noticed that these are public api so we should document them too.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.h:
```diff
@@ -0,0 +1,153 @@
+  *THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  *(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
```
`*` are weirdly located here

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.h:
```diff
@@ -0,0 +1,153 @@
+  *@WPE_SETTINGS_ERROR_ALREADY_REGISTERED: Key has already been registered
+ *
+  *#WPESettings errors
+ */
```
And here

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.h:
```diff
@@ -0,0 +1,153 @@
+#define WPE_TYPE_SETTINGS (wpe_settings_get_type())
+WPE_API G_DECLARE_FINAL_TYPE(WPESettings, wpe_settings, WPE, SETTINGS, GObject)
+
+WPE_API WPESettings    *wpe_settings_get_default                    ();
```
There are extra spaces between return type and function name and between function name and params here in all the cases.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+    GRefPtr<GVariant> setValue;
+    GRefPtr<GVariant> defaultValue;
+    GUniquePtr<GVariantType> type;
+    bool setByApplication { false };
```
I'm sorry I couldn't do a proper review yesterday, just checked the smart pointer issues with the floating references. So, my point regarding this was that, even when the intention of the api is to allow apps to override settings preventing the platforms from changing them, we can't avoid platforms passing TRUE or even apps passing FALSE meaning, only change the setting is not explicitly set by the platform. But I see we can't make it generic this way, because then platform could not overwrite their own settings. So, after thinking about it, I think we could use an enum parameter, WPESettingsSource or WPESettingsChangeSource, or similar, with values APPLICATION/PLATFORM or something like that, and document that platform implementations should pass PLATFORM as the source parameter. What do you think?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+ * The key must begin with `/wpe-platform/` and not end with `/`. e.g. `/wpe-platform/fonts/hinting`.
+ *
+ * It is an error to call this function with a key that has already been registered.
+ */
```
We should also document that if default_value has a floating ref it's consumed.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+ * wpe_settings_set_value:
+ * @settings: a #WPESettings
+ * @key: the key to set
+ * @value: (transfer floating) (nullable): the value to set or %NULL
```
transfer floating is like transfer none, we are not using it in register method for the default value. I think it's confusing, but if you want to keep it, add it to register too.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+ * Setting @value to %NULL will reset it to the default.
+ *
+ * Returns: %TRUE if the value was set successfully, %FALSE otherwise.
+ */
```
And we should document that if value has a floating ref, it's consumed.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+ *
+ * Setting @value to %NULL will reset it to the default.
+ *
+ * Returns: %TRUE if the value was set successfully, %FALSE otherwise.
```
In this case I would say returns %FALSE in case of error, or %TRUE otherwise or something like that, because when override is FALSE and the value has been set before we don't really set anything, but TRUE is returned.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+        iter->value.setValue = nullptr;
+        iter->value.setByApplication = overridePlatform;
+        return TRUE;
+    }
```
This is actually changing the setting, by resetting it to the default value, so I guess we should emit the changed signal. Adding a test case for this would be great too.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
```diff
@@ -0,0 +1,504 @@
+        return FALSE;
+    }
+
+    iter->value.setValue = g_variant_ref_sink(value);
```
```suggestion
    iter->value.setValue = value;
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WPEPlatform/TestSettings.cpp:
```diff
@@ -0,0 +1,231 @@
+    wpe_settings_set_value(settings.get(), "/wpe-platform/a-key", someValue.get(), FALSE, nullptr);
+    wpe_settings_set_value(settings.get(), "/wpe-platform/another-key", anotherValue.get(), FALSE, nullptr);
+    g_signal_handler_disconnect(settings.get(), handler);
+    g_assert_cmpint(count, ==, 1);
```
Why is it emitted in this case? The same value is set, no? hmm, I guess we don't compare to the default value?

TingPing replied:
Changed to only emit when new value doesn't equal old.

TingPing committed:
[WPE] WPE Platform: Add WPESettings API
https://bugs.webkit.org/show_bug.cgi?id=280618

Reviewed by Carlos Garcia Campos.

This adds an API to store generic settings similar to GSettings.

Each setting has a known type and default value.

These settings can be saved and loaded from key files.

* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_settings):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/WPESettings.cpp: Added.
(SettingEntry::SettingEntry):
(SettingEntry::value const):
(wpe_settings_get_default):
(makeKeyPath):
(wpe_settings_register):
(wpe_settings_load_from_keyfile):
(wpe_settings_save_to_keyfile):
(wpe_settings_set_value):
(wpe_settings_get_value):
* Source/WebKit/WPEPlatform/wpe/WPESettings.h: Added.
* Tools/Scripts/run-wpe-tests:
(WPETestRunner):
(WPETestRunner.is_glib_test):
* Tools/TestWebKitAPI/Tests/WPEPlatform/TestSettings.cpp: Added.
(testSettingsRegistration):
(testSettingsSaveKeyFiles):
(testSettingsLoadKeyFiles):
(testSettingsChangedSignal):
(testSettingsHelperAPI):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

Canonical link: https://commits.webkit.org/285702@main
